### PR TITLE
ARROW-1727: [Format] Expand Arrow streaming format to permit deltas / additions to existing dictionaries

### PR DESCRIPTION
**Pull request by TheNeuralBit**
Add an `isDelta` flag to the `DictionaryBatch` to allow for dictionary modifications mid-stream, update documentation.

### CODE REVIEW

File: format/IPC.md [hunk 1]
```diff
@@ -67,15 +67,18 @@ We provide a streaming format for record batches. It is presented as a sequence
...
 
 ```
 <SCHEMA>
 <DICTIONARY 0>
 ...
-<DICTIONARY k - 1>
 <RECORD BATCH 0>
 ...
+<DICTIONARY k - 1>
+...
```
Comment (wesm) [inline format/IPC.md]: This is slightly misleading. I believe that all dictionaries need to appear before the first record batch, but then arbitrary delta batches may appear after the first record batch

File: format/Message.fbs [hunk 2]
```diff
@@ -61,16 +61,17 @@ table RecordBatch {
...
-/// There is one vector / column per dictionary
-///
+/// There is one vector / column per dictionary, but that vector / column
+/// may be spread across multiple dictionary batches by using the isDelta
+/// flag
 
 table DictionaryBatch {
   id: long;
   data: RecordBatch;
+  isDelta: bool = false;
```
Comment (wesm) [inline format/Message.fbs]: Can you add a comment here "If isDelta is true, the values in the dictionary are to be appended to an dictionary with the indicated id"

File: format/IPC.md [hunk 3]
```diff
@@ -189,6 +197,10 @@ in the schema, so that dictionaries can even be used for multiple fields. See
 the [Physical Layout][4] document for more about the semantics of
 dictionary-encoded data.
 
+The dictionary `isDelta` flag allows dictionary batches to be modified mid-stream.
+A dictionary batch with `isDelta` set indicates that its vector should be
+concatenated with those of any previous batches with the same `id`.
```
Comment (wesm) [inline format/IPC.md]: For the avoidance of confusion, we might want to include a small pseudocode example of a stream with a delta dictionary batch

**Review by wesm — COMMENTED** [review]
Thanks @TheNeuralBit! Left a few minor comments

File: format/IPC.md (hunk 3 above)
Comment (TheNeuralBit) [reply format/IPC.md]: Would something like the example in my initial email work?

File: format/IPC.md (hunk 1 above)
Comment (TheNeuralBit) [reply format/IPC.md]: Yeah thats fair, I can tweak it to make it clear that any dictionaries after the first record batch should be modifying the originals.

**Review by wesm — APPROVED** [review]
+1

**Comment by wesm** [comment]
thanks @TheNeuralBit!

**Pull request closed without merging**